elyra/pipeline/catalog_connector.py

```diff
@@ -27,13 +27,116 @@
+
+
+class EntryData(object):
+    """
+    An object representing the data retrieved from a single entry of a catalog, which,
+    at minimum, includes the string definition of the corresponding component(s)
+    """
+    definition: str = None
+
+    def __init__(self, definition: str, **kwargs):
+        self.definition = definition
+
+
+class AirflowEntryData(EntryData):
+    """
+    An Airflow-specific EntryData object that includes the fully-qualified package
+    name (excluding class name) that represents the definition file.
+    """
+    package_name: str = None
+
+    def __init__(self, definition: str, **kwargs):
+        super().__init__(definition, **kwargs)
+        self.package_name = kwargs.get('package_name')
+
+
+class KfpEntryData(EntryData):
+    """
+    A KFP-specific EntryData object
+    """
+    pass
+
+
+class CatalogEntry(object):
+    """
+    An object corresponding to a single entry of a component catalog, which has a
+    unique id, a string definition, a dict of identifying key-value pairs, and
+    other associated metadata.
+    """
+    id: str
+    entry_data: EntryData
+    entry_reference: Any
+    catalog_type: str
+    runtime_type: RuntimeProcessorType
+    categories: List[str]
+
+    def __init__(self,
+                 entry_data: EntryData,
+                 entry_reference: Any,
+                 catalog_instance: Metadata,
+                 hash_keys: List[str]):
+        self.entry_data = entry_data
+        self.entry_reference = entry_reference
+        self.catalog_type = catalog_instance.schema_name
+        self.runtime_type = catalog_instance.runtime_type.name  # noqa
+        self.categories = catalog_instance.metadata.get("categories", [])
+
+        self.id = self.compute_unique_id(hash_keys)
+
+    def compute_unique_id(self, hash_keys: List[str]) -> str:
+        """
+        Computes a unique id for the given component based on the schema name of the
+        catalog connector type and any information specific to that component-catalog-type
+        combination as given in hash_keys.
+
+        :param hash_keys: the list of keys (present in the catalog_entry_data dict)
+            whose values will be used to construct the hash
+
+        :returns: a unique component id of the form '<catalog-type>:<hash_of_entry_info>'
+        """
+        hash_str = ""
+        for key in hash_keys:
+            if not self.entry_reference.get(key):
+                # Catalog entry does not have key - build hash without it
+                continue
+            hash_str = hash_str + str(self.entry_reference[key]) + ":"
+        hash_str = hash_str[:-1]
+
+        # Use only the first 12 characters of the resulting hash
+        hash_digest = f"{hashlib.sha256(hash_str.encode()).hexdigest()[:12]}"
+        return f"{self.catalog_type}:{hash_digest}"
+
+    def get_component(self, id: str, name: str, description: str, properties: List[ComponentParameter]) -> Component:
```

Comment:
We should probably have a help-string for this helper method.

Reply:
ah yes good catch. I had one here once upon a time 😆